### PR TITLE
FullCalendar v6 is added

### PR DESCRIPTION
**Pull request by solomax**

### CODE REVIEW

**Comment by solomax** [comment]
@reckart please take a look :)

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarAdapter.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import java.time.LocalDateTime;
+
+import org.apache.wicket.ajax.AjaxRequestTarget;
+
+/**
+ * Adapter class for {@link ICalendarListener}
+ *
+ * @author Sebastien Briquet - sebfz1
+ */
+public class CalendarAdapter implements ICalendarListener
```
Comment (reckart) [inline wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarAdapter.java:29]: Since we have default methods in interfaces these days in Java, such adapter classes should no longer be necessary.

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarEvent.java [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+		{
+			options.set("color", Options.asString(this.color));
+		}
+
+		if (this.backgroundColor != null)
+		{
+			options.set("backgroundColor", Options.asString(this.backgroundColor));
+		}
+
+		if (this.borderColor != null)
```
Comment (reckart) [inline wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarEvent.java:279]: The code could probably be much reduced by having an alternative option.set method (maybe setUnlessNull).

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarModel.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+	private LocalDate start;
+	private LocalDate end;
+
+	/**
+	 * Constructor
+	 */
+	public CalendarModel()
+	{
+		this.start = null;
```
Comment (reckart) [inline wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarModel.java:43]: This null assignment is redundant. If there is no constructor with arguments, then a no-args constructor is optional.

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarModelBehavior.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+			this.setEndDate(this.model, parse(end));
+
+			List<? extends CalendarEvent> list = this.model.getObject(); // calls load()
+
+			if (list != null)
+			{
+				int count = 0;
+				for (CalendarEvent event : list)
+				{
+					if (this.model instanceof ICalendarVisitor)
```
Comment (reckart) [inline wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarModelBehavior.java:93]:
```
if (this.model instanceof ICalendarVisitor visitor) {
  event.accept(visitor);
}
```

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarView.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	timeGridDay,
+
+	multiMonthYear,
+
+	listDay,
+	listWeek,
+	listMonth,
+	listYear;
+
+	private static final Logger LOG = LoggerFactory.getLogger(CalendarView.class);
```
Comment (reckart) [inline wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarView.java:43]: `LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass())`?

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/ICalendarListener.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+import org.apache.wicket.ajax.AjaxRequestTarget;
+import org.apache.wicket.util.io.IClusterable;
+
+/**
+ * Event listener shared by the {@link Calendar} widget and the {@link CalendarBehavior}
+ *
+ * @author Sebastien Briquet - sebfz1
+ *
+ */
+public interface ICalendarListener extends IClusterable
```
Comment (reckart) [inline wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/ICalendarListener.java:31]: Best add no-op default implementations here.

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/ICalendarVisitor.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+
+/**
+ * Provides the ability for a {@link CalendarEvent} to be visited by a {@link CalendarModel}, after events have been retrieved by the {@link CalendarModelBehavior} (after {@link CalendarModel#load()}) <br>
+ * A typical use is to set the css class-name ({@link CalendarEvent#setClassName(String)}) depending on another property.
+ *
+ * @author Sebastien Briquet - sebfz1
+ *
+ */
+public interface ICalendarVisitor
```
Comment (reckart) [inline wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/ICalendarVisitor.java:27]: Should the interface be marked as functional? https://docs.oracle.com/javase/8/docs/api/java/lang/FunctionalInterface.html

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/resource/CalendarJavaScriptResourceReference.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+ * Provides the resource reference for the fullcalendar javascript library.
+ *
+ * @author Sebastien Briquet - sebfz1
+ *
+ */
+public class CalendarJavaScriptResourceReference extends WebjarsJavaScriptResourceReference
+{
+	private static final long serialVersionUID = 1L;
+
+	private static class Holder {
```
Comment (reckart) [inline wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/resource/CalendarJavaScriptResourceReference.java:31]: What is the benefit of the inner holder class vs. having the instance field be a static private final field of the outer resource reference class?

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/Calendar.java [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+	}
+
+	@Override
+	public void onConfigure()
+	{
+		super.onConfigure();
+
+		// builds sources //
+		StringBuilder sourceBuilder = new StringBuilder();
+		sourceBuilder.append(String.format("{ url: '%s' }", Calendar.this.modelBehavior.getCallbackUrl()));
```
Comment (martin-g) [inline wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/Calendar.java:200]: nit: there is no need of `Calendar.` in `Calendar.this`

File: wicket-jquery-ui-parent/wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/component/DemoCalendar6Dialog.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+				Boolean allday = this.getModelObject();
+				startDateTimePicker.setTimePickerEnabled(!allday);
+				endDateTimePicker.setTimePickerEnabled(!allday);
+			}
+
+			@Override
+			protected void onUpdate(AjaxRequestTarget target)
+			{
+				Boolean allday = this.getModelObject();
```
Comment (reckart) [inline wicket-jquery-ui-parent/wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/component/DemoCalendar6Dialog.java:79]: Can use `boolean` (or `var`) here. The `null` case is not handled later anyway.

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/Calendar.java [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+		if (Calendar.this.sources != null)
+		{
+			for (EventSource source : Calendar.this.sources)
+			{
+				sourceBuilder.append(", ");
+				sourceBuilder.append(source.toString());
+			}
+		}
+
+		behavior.setOption("eventSources", String.format("[%s]", sourceBuilder.toString()));
```
Comment (martin-g) [inline wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/Calendar.java:211]:
```suggestion
		behavior.setOption("eventSources", "[" + sourceBuilder.toString() + "]");

```
Shorter code and faster at runtime

File: wicket-jquery-ui-parent/wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/data/DemoCalendar6Model.java [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+import com.googlecode.wicket.jquery.ui.samples.data.dao.Calendar6DAO;
+
+public class DemoCalendar6Model extends CalendarModel implements ICalendarVisitor
+{
+	private static final long serialVersionUID = 1L;
+
+	@Override
+	protected List<DemoCalendar6Event> load()
+	{
+		 LocalDate start = this.getStart();
```
Comment (reckart) [inline wicket-jquery-ui-parent/wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/data/DemoCalendar6Model.java:18]: I would inline these variables.

File: wicket-jquery-ui-parent/wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/jqueryui/calendar6/CustomCalendar6Page.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+	private CalendarModel newCalendarModel()
+	{
+		return new CalendarModel() {
+
+			private static final long serialVersionUID = 1L;
+
+			@Override
+			protected List<CalendarEvent> load()
+			{
+				// Loads events from the data access layer, using:
```
Comment (reckart) [inline wicket-jquery-ui-parent/wicket-jquery-ui-samples/src/main/java/com/googlecode/wicket/jquery/ui/samples/jqueryui/calendar6/CustomCalendar6Page.java:48]: I don't understand the comment. Is it useful?

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarBehavior.java [hunk 14]
```diff
@@ -0,0 +1,1005 @@
...
+	private static final long serialVersionUID = 1L;
+	public static final String METHOD = "fullCalendar";
+	private final Options options;
+	private final String selector;
+
+	/** event listener */
+	private final ICalendarListener listener;
+
+	/** date range-select behavior */
+	private JQueryAjaxBehavior onSelectAjaxBehavior = null;
```
Comment (martin-g) [inline wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarBehavior.java:60]:
```suggestion
	private JQueryAjaxBehavior onSelectAjaxBehavior;

```

**Review by martin-g — APPROVED** [review]
Few nits only.

File: wicket-jquery-ui-parent/wicket-kendo-ui/pom.xml [hunk 15]
```diff
@@ -13,8 +13,8 @@
 	<artifactId>wicket-kendo-ui</artifactId>
 	<packaging>bundle</packaging>
 
-    <properties>
-	    <bundle.namespace>com.googlecode.wicket.kendo.ui</bundle.namespace>
+	<properties>
+		<bundle.namespace>com.googlecode.wicket.kendo.ui</bundle.namespace>
```
Comment (reckart) [inline wicket-jquery-ui-parent/wicket-kendo-ui/pom.xml:17]: This change seems unrelated. I'm also a bit confused because the master branch says `org.wicketstuff.kendo.ui` here.

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/pom.xml [hunk 16]
```diff
@@ -13,8 +13,8 @@
 	<artifactId>wicket-jquery-ui-calendar</artifactId>
 	<packaging>bundle</packaging>
 
-    <properties>
-	    <bundle.namespace>com.googlecode.wicket.jquery.ui.calendar</bundle.namespace>
+	<properties>
+		<bundle.namespace>com.googlecode.wicket.jquery.ui.calendar</bundle.namespace>
```
Comment (reckart) [inline wicket-jquery-ui-parent/wicket-jquery-ui-calendar/pom.xml:17]: I don't understand why this change shows up like this here. The master branch says `org.wicketstuff.jquery.ui.calendar` here.

**Review by reckart — COMMENTED** [review]
It seems that all files in the samples part of the PR are lacking a license header.

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/Calendar.java (hunk 11 above)
Comment (solomax) [reply wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/Calendar.java]: fixed

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarBehavior.java (hunk 14 above)
Comment (solomax) [reply wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarBehavior.java]: fixed

**Comment by solomax** [comment]
> It seems that all files in the samples part of the PR are lacking a license header.

fixed

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarModelBehavior.java (hunk 4 above)
Comment (solomax) [reply wicket-jquery-ui-parent/wicket-jquery-ui-calendar/src/main/java/com/googlecode/wicket/jquery/ui/calendar6/CalendarModelBehavior.java:93]: this is impossible for java 11 :) (will update in `10.x`)

File: wicket-jquery-ui-parent/wicket-jquery-ui-calendar/pom.xml (hunk 16 above)
Comment (solomax) [reply wicket-jquery-ui-parent/wicket-jquery-ui-calendar/pom.xml:17]: spaces -> tabs :))) the code mixes these two :(

**Comment by solomax** [comment]
@reckart I'm ready to merge :)
